### DISCUSSION TITLE
Custom aggregation functions

### DISCUSSION THREAD

**Discussion by imatveev** [oguimbal/pg-mem#111] [Q&A] [answered]
Currently, there is a way to create custom functions through `registerFunction`. Together with that - when I try to create a custom function in aggregated queries it throws  🔨 Not supported 🔨 : aggregation function.
Is there a way to define new aggregation functions by yourself?

**Comment by oguimbal** [comment, marked as answer]
Not yet, no.

I have started a refactoring that might lead to making this possible, though.

**Reply by imatveev** [reply]
Got it, thanks! I will use some workaround till then, and follow new versions.
Thanks for your efforts!

**Reply by imatveev** [reply]
@oguimbal Would you mind me trying to contribute `array_agg` function?

**Reply by oguimbal** [reply]
Of course not :)

[Here is some inspiration](https://github.com/oguimbal/pg-mem/blob/f3c01089aa540e8c7938424456f2478523a6f110/src/transforms/aggregation.ts#L290-L302)

That said, you will understand pretty soon that the current pg-mem implementation does not support controling the order  of aggregation via  `array_agg(xxx order by yyy)`. That was not an issue for the aggregations I implemented so far, but it will for `arrag_agg`  ☹
... that is the point of the incoming refactoring, though

**Reply by imatveev** [reply]
@oguimbal During my tests I noticed that ordering is not supported not only inside aggregated functions but in aggregated queries at all. Something like this:
```
    it('supports ordering with aggregated queries', () => {
        expect(many(`create table example(a int, b int);
                    insert into example values (1, 1), (1, 1), (3, 3), (3, 3), (2, 2), (2, 2);
                    select a, max(b) as b from example group by a order by a ASC`))
            .to.deep.equal([{ a: 1, b: 1 }, { a: 2, b: 2 }, { a: 3, b: 3 }]);
    });
 // Throws an error, the returned order is [{ a: 1, b: 1 }, { a: 3, b: 3 }, { a: 2, b: 2 }]
```
Is it what you meant also, or this is a separate issue?

**Reply by oguimbal** [reply]
Yes, the current execution engine is a bit broken when it comes to aggregations & nested queries.

Working on it :) Thanks for your unit test, I'm adding it to the suite !

**Reply by BasitAli** [reply]
It's been a long time since this was last discussed, but do we have a way now to add custom aggregation functions?

**Reply by Thiamath** [reply]
Hey! Just sneaky nudging this topic!
I found a small problem when using `array_agg` in a view in my schema. It was supposed to aggregate `uuid` types and so return `uuid[]` but it seems to be returning `integer[]`.
Any ideas?
